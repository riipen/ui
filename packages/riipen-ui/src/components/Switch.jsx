import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";

const Switch = props => {
  const {
    checked,
    classes,
    color,
    disabled,
    id,
    onChange,
    required,
    size,
    value,
    ...other
  } = props;

  const theme = React.useContext(ThemeContext);

  return (
    <span className={clsx(classes)}>
      <input
        checked={checked}
        className={clsx(color, other.readOnly ? "read-only" : null, size)}
        disabled={disabled}
        id={id}
        onChange={onChange}
        required={required}
        type="checkbox"
        value={value}
        {...other}
      />
      <style jsx>{`
        input {
          background-color: ${theme.palette.grey[400]};
          cursor: pointer;
          outline: none;
          position: relative;
          transition: ${theme.transitions.create(["background-color"], {
            duration: theme.transitions.duration.standard
          })};
          -webkit-appearance: none;
        }
        input:before {
          background-color: ${theme.palette.grey[600]};
          content: "";
          left: 0;
          position: absolute;
          top: 0;
          transition: ${theme.transitions.create(
            ["transform", "background-color"],
            {
              duration: theme.transitions.duration.standard
            }
          )};
        }
        input:hover:before {
          box-shadow: 0px 0px 7px 1px ${theme.palette.grey[500]};
        }

        .read-only {
          cursor: initial;
          pointer-events: none;
        }

        input:disabled,
        input:checked:disabled {
          background-color: ${theme.palette.grey[200]};
          pointer-events: none;
        }
        input:disabled:before,
        input:disabled:checked:before {
          background-color: ${theme.palette.disabled};
        }

        .medium {
          border-radius: 12px;
          height: 20px;
          width: 40px;
        }
        .medium:before {
          border-radius: 50%;
          height: 20px;
          width: 20px;
        }
        .medium:checked:before {
          transform: translate(20px);
        }

        .small {
          border-radius: 10px;
          height: 15px;
          width: 30px;
        }
        .small:before {
          border-radius: 50%;
          height: 15px;
          width: 15px;
        }
        .small:checked:before {
          transform: translate(15px);
        }

        .primary:checked {
          background-color: ${theme.palette.primary.light};
        }
        .primary:checked:before {
          background-color: ${theme.palette.primary.dark};
        }

        .secondary:checked {
          background-color: ${theme.palette.secondary.light};
        }
        .secondary:checked:before {
          background-color: ${theme.palette.secondary.dark};
        }

        .tertiary:checked {
          background-color: ${theme.palette.tertiary.light};
        }
        .tertiary:checked:before {
          background-color: ${theme.palette.tertiary.dark};
        }

        .negative:checked {
          background-color: ${theme.palette.negative.light};
        }
        .negative:checked:before {
          background-color: ${theme.palette.negative.dark};
        }

        .positive:checked {
          background-color: ${theme.palette.positive.light};
        }
        .positive:checked:before {
          background-color: ${theme.palette.positive.dark};
        }
      `}</style>
    </span>
  );
};

Switch.propTypes = {
  /**
   * If true, the input is checked.
   */
  checked: PropTypes.bool,

  /**
   * An array of custom CSS classes to apply.
   */
  classes: PropTypes.array,

  /**
   * The color of the switch.
   */
  color: PropTypes.oneOf([
    "primary",
    "secondary",
    "tertiary",
    "positive",
    "negative"
  ]),

  /**
   * If true, the switch will be disabled.
   */
  disabled: PropTypes.bool,

  /**
   * The id of the input element.
   */
  id: PropTypes.string,

  /**
   * Callback fired when the state is changed.
   *
   * @param {object} event The event source of the callback.
   * You can pull out the new value by accessing event.target.value (string). You can pull out the new checked state by accessing event.target.checked (boolean).
   */
  onChange: PropTypes.func,

  /**
   * If true, the input element will be required.
   */
  required: PropTypes.bool,

  /**
   * The size of the switch. small is equivalent to the dense switch styling.
   */
  size: PropTypes.oneOf(["small", "medium"]),

  /**
   * @ignore
   * The theme context object
   */
  theme: PropTypes.object,

  /**
   * The value of the component. The DOM API casts this to a string. The browser uses "on" as the default value.
   */
  value: PropTypes.any
};

Switch.defaultProps = {
  checked: false,
  classes: [],
  color: "secondary",
  size: "medium"
};

Switch.displayName = "Switch";

export default Switch;
