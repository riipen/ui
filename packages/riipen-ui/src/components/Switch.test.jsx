import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Switch from "./Switch";

describe("<Switch>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Switch />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Switch />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Switch />);

      const component = wrapper.find("Switch");
      expect(component.props().checked).toEqual(false);
      expect(component.props().classes).toEqual([]);
      expect(component.props().color).toEqual("secondary");
      expect(component.props().size).toEqual("medium");
    });
  });

  describe("checked prop", () => {
    it("passes given checked value to input element", () => {
      const checked = true;

      const wrapper = mount(<Switch checked={checked} />);

      expect(wrapper.find("input").props().checked).toEqual(checked);
    });
  });

  describe("classes prop", () => {
    it("applies given classes to the root node", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Switch classes={classes} />);

      expect(
        wrapper
          .find("Switch")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });
  });

  describe("color prop", () => {
    it("applies valid custom color class name", () => {
      const color = "positive";

      const wrapper = mount(<Switch color={color} />);

      expect(wrapper.find("input").hasClass(color)).toEqual(true);
    });

    it("gives an error given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const color = "default";

      mount(<Switch color={color} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("disabled prop", () => {
    it("passes given disabled value to input element", () => {
      const disabled = true;

      const wrapper = mount(<Switch disabled={disabled} />);

      expect(wrapper.find("input").props().disabled).toEqual(true);
    });
  });

  describe("id", () => {
    it("passes given id value to input element", () => {
      const id = "ID";

      const wrapper = mount(<Switch id={id} />);

      expect(wrapper.find("input").props().id).toEqual(id);
    });
  });

  describe("onChange prop", () => {
    it("invokes onChange handler on change event", () => {
      const onChange = jest.fn();
      const wrapper = mount(<Switch onChange={onChange} checked />);

      wrapper.find("input").simulate("change");

      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  describe("required prop", () => {
    it("passes given required value to input element", () => {
      const required = true;

      const wrapper = mount(<Switch required={required} />);

      expect(wrapper.find("input").props().required).toEqual(required);
    });
  });

  describe("size prop", () => {
    it("applies valid custom size class name to input element", () => {
      const size = "small";

      const wrapper = mount(<Switch size={size} />);

      expect(wrapper.find("input").hasClass(size)).toEqual(true);
    });

    it("gives an error given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const size = "large";

      mount(<Switch size={size} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("value prop", () => {
    it("passes given value to input element", () => {
      const value = "pizza";

      const wrapper = mount(<Switch value={value} />);

      expect(wrapper.find("input").props().value).toEqual(value);
    });
  });
});
