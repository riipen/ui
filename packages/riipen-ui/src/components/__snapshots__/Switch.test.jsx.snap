// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch> renders correct snapshot 1`] = `
<Switch
  classes={Array []}
  color="secondary"
  size="medium"
>
  <span
    className=""
  >
    <input
      className="secondary medium"
      onChange={[Function]}
      type="checkbox"
    />
    <style>
      
        input {
          -webkit-appearance: none;
          outline: none;
          background-color: #cccbcb;
          position: relative;
          transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
        }
        input:before {
          content: "";
          position: absolute;
          top: 0;
          left: 0;
          background-color: #747474;
          transition: transform 300ms cubic-bezier(0.4, 0, 0.2, 1), background-color 300ms cubic-bezier(0.4, 0, 0.2, 1);
        }
        input:hover:before {
          box-shadow: 0px 0px 15px 3px #9a9a9a;
        }

        input:disabled, input:checked:disabled {
          background-color: #747474;
          pointer-events: none;
        }
        input:disabled:before, input:disabled:checked:before {
          background-color: #424242;
        }

        .medium {
          width: 40px;
          height: 20px;
          border-radius: 12px;
        }
        .medium:before {
          height: 20px;
          width: 20px;
          border-radius: 50%;
        }
        .medium:checked:before {
          transform: translate(20px);
        }

        .small {
          width: 30px;
          height: 15px;
          border-radius: 10px;
        }
        .small:before {
          height: 15px;
          width: 15px;
          border-radius: 50%;
        }
        .small:checked:before {
          transform: translate(15px);
        }
        
        .primary:checked {
          background-color: #2c8ddc;
        }
        .primary:checked:before {
          background-color: #254364;
        }

        .secondary:checked {
          background-color: #f5bf95;
        }
        .secondary:checked:before {
          background-color: #ed7321;
        }

        .tertiary:checked {
          background-color: #78becc;
        }
        .tertiary:checked:before {
          background-color: #41978a;
        }

        .negative:checked {
          background-color: #ea7e7e;
        }
        .negative:checked:before {
          background-color: #ba2222;
        }

        .positive:checked {
          background-color: #78dbac;
        }
        .positive:checked:before {
          background-color: #1b965c;
        }
      
    </style>
  </span>
</Switch>
`;
